clang/lib/Headers/cheriintrin.h

```diff
@@ -0,0 +1,130 @@
+
+/* Bounds setting intrinsics: */
+#define cheri_round_representable_length(x)                                    \
+  __builtin_cheri_round_representable_length(x)
+#define cheri_representable_alignment_mask(x)                                  \
+  __builtin_cheri_representable_alignment_mask(x)
+#define cheri_set_bounds(x, y) __builtin_cheri_bounds_set((x), (y))
+#define cheri_set_bounds_exact(x, y) __builtin_cheri_bounds_set_exact((x), (y))
+
+/* Object types, sealing and unsealing: */
+typedef long cheri_otype_t;
+#if defined(__mips__) || defined(__riscv)
+/* CHERI-MIPS and CHERI-RISC-V use negative numbers for hardware-interpreted
+ * otypes */
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)-1)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)-2)
+#else
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)0)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)1)
+#endif
+#define cheri_get_type(x) __builtin_cheri_type_get(x)
+#define cheri_is_sealed(x) __builtin_cheri_sealed_get(x)
+#define cheri_is_sentry(x) (__builtin_cheri_type_get(x) == CHERI_OTYPE_SENTRY)
+#define cheri_is_unsealed(x) (!__builtin_cheri_sealed_get(x))
+/* TODO: builtins for indirect sentries */
+#define cheri_seal_entry(x) __builtin_cheri_seal_entry(x)
+#define cheri_seal(x, y) __builtin_cheri_seal((x), (y))
+#define cheri_unseal(x, y) __builtin_cheri_unseal((x), (y))
+
+/* Reconstruct capabilities from raw data: */
+#define cheri_build_cap(x, y) __builtin_cheri_cap_build((x), (y))
+#define cheri_conditional_seal(x, y) __builtin_cheri_conditional_seal((x), (y))
+#define cheri_copy_type(x, y) __builtin_cheri_cap_type_copy((x), (y))
+
+/* Capability permissions: */
+typedef enum __attribute__((flag_enum, enum_extensibility(open))) {
+  CHERI_PERM_GLOBAL = __CHERI_CAP_PERMISSION_GLOBAL__,
+  CHERI_PERM_EXECUTE = __CHERI_CAP_PERMISSION_PERMIT_EXECUTE__,
+  CHERI_PERM_LOAD = __CHERI_CAP_PERMISSION_PERMIT_LOAD__,
+  CHERI_PERM_STORE = __CHERI_CAP_PERMISSION_PERMIT_STORE__,
+  CHERI_PERM_LOAD_CAP = __CHERI_CAP_PERMISSION_PERMIT_LOAD_CAPABILITY__,
+  CHERI_PERM_STORE_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_CAPABILITY__,
+  CHERI_PERM_STORE_LOCAL_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_LOCAL__,
+  CHERI_PERM_SEAL = __CHERI_CAP_PERMISSION_PERMIT_SEAL__,
+  CHERI_PERM_CCALL = __CHERI_CAP_PERMISSION_PERMIT_CCALL__,
+  CHERI_PERM_UNSEAL = __CHERI_CAP_PERMISSION_PERMIT_UNSEAL__,
+  CHERI_PERM_SYSTEM_REGS = __CHERI_CAP_PERMISSION_ACCESS_SYSTEM_REGISTERS__,
+  /* TODO: architecture-dependent permissions */
+} cheri_perms_t;
+#define cheri_get_perms(x) ((cheri_perms_t)(__builtin_cheri_perms_get(x)))
+#define cheri_and_perms(x, y) __builtin_cheri_perms_and((x), (__SIZE_TYPE__)(y))
+#define cheri_clear_perms(x, y)                                                \
+  __builtin_cheri_perms_and((x), ~(__SIZE_TYPE__)(y))
+
+/* Accessors for capability registers. Currently exposes DDC and PCC. */
+#define cheri_get_ddc() __builtin_cheri_global_data_get()
```

Comment:
Do we not want a cheri_set_ddc?

Reply:
Could be useful, but we don't currently have a way of expressing that other than inline assembly.

Reply:
Hm, yeah, and I doubt builtins can have memory clobbers. We _could_ define the inline assembly here, but it's a pretty rare thing to want to do, neither hybrid nor purecap do it in normal use.